src/vm/contexts.rs

```diff
@@ -559,7 +589,14 @@ impl <'a,'b> Environment <'a,'b> {
 
         let args = args?;
 
-        self.execute_function_as_transaction(&func, &args, Some(&contract.contract_context)) 
+        let func_identifier = func.get_identifier();
+        if self.call_stack.contains(&func_identifier) {
```

Comment:
Is this ever true?  Isn't the type-checker supposed to ensure that there are no reference cycles before the contract is instantiated?

Reply:
We run the type-checker only once - when you publish a contract.
When publishing a contract, the only thing the type-checker knows, is the arguments and the type returned by the dynamic call.
When executing a contract-call transaction we can assemble the puzzle. This piece of code is aborting the execution when reentrancy is detected (test case available).

Reply:
Ah, I see.  Thanks!